### PR TITLE
feat: support different decimals value between chains

### PR DESCRIPTION
**Pull request by nmlinaric**
## Description
This adds allows handlers to convert decimal values when bridging between chains that have different decimal places.

## Related Issue Or Context
Closes: #59 

## How Has This Been Tested? Testing details.
Unit tests, e2e tests

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation

## Checklist:
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have ensured that all acceptance criteria (or expected behavior) from issue are met
- [ ] I have updated the documentation locally and in chainbridge-docs.
- [x] I have added tests to cover my changes.
- [x] I have ensured that all the checks are passing and green, I've signed the CLA bot

### CODE REVIEW

**Review by P1sar — COMMENTED** [review]
This looks like a pretty breaking changes, maybe we can have this implemented as a separate handler in order to not increase general ERC20Handler compleixty?

**Comment by P1sar** [comment]
> 

Deploying one contract is only 1-time cost increase. Moreover, as for me, this was the initial architectural design idea to have Handler separate so we do not bring breaking changes to the existing handler but implement new ones for new cases. 
cc @lastperson @mpetrun5

File: contracts/handlers/ERC20Handler.sol [hunk 1]
```diff
@@ -115,4 +115,19 @@ contract ERC20Handler is IHandler, ERCHandlerHelpers, ERC20Safe {
...
     }
+
+    /**
+        @notice Converts number of decimals from destination to origin chain.
+        @param tokenAddress Address of contract to be used when making or executing deposits.
+        @param amount Decimals value to be set for {contractAddress}.
+    */
+    function convertBalance(address tokenAddress, uint256 amount) internal returns(uint256) {
+        Decimals memory decimals = _decimals[tokenAddress];
+        require(decimals.srcDecimals != 0 && decimals.destDecimals != 0, "Invalid decimals");
```
Comment (lastperson) [inline contracts/handlers/ERC20Handler.sol:126]: Zero decimals should be considered valid, it doesn't break anything here.

File: contracts/handlers/ERC20Handler.sol [hunk 2]
```diff
@@ -115,4 +115,19 @@ contract ERC20Handler is IHandler, ERCHandlerHelpers, ERC20Safe {
     function setResource(bytes32 resourceID, address contractAddress, bytes calldata args) external onlyBridge {
         _setResource(resourceID, contractAddress);
     }
+
+    /**
+        @notice Converts number of decimals from destination to origin chain.
+        @param tokenAddress Address of contract to be used when making or executing deposits.
+        @param amount Decimals value to be set for {contractAddress}.
```
Comment (lastperson) [inline contracts/handlers/ERC20Handler.sol:122]: Comment seems unrelated.

File: contracts/handlers/ERCHandlerHelpers.sol [hunk 3]
```diff
@@ -24,6 +24,14 @@ contract ERCHandlerHelpers is IERCHandler {
     // token contract address => is burnable
     mapping (address => bool) public _burnList;
 
+    // token contract address => decimals
+    mapping (address => Decimals) public _decimals;
```
Comment (lastperson) [inline contracts/handlers/ERCHandlerHelpers.sol:28]: TODO: create a task to pack _contractWhitelist, _burnList and _decimals into a single mapping.

File: contracts/Bridge.sol [hunk 4]
```diff
@@ -159,6 +159,18 @@ contract Bridge is Pausable, Context, EIP712 {
...
+      /**
+        @notice Sets source and destination number of decimals for handler contract and provided token,
+        which is mapped in {_decimals} in ERCHandlerHelpers.
+        @notice Only callable by address that has the right to call the specific function.
+        @param handlerAddress Address of handler resource will be set for.
+        @param tokenAddress Address of contract to be called when a deposit is made and a deposited is executed.
+     */
+    function adminSetDecimals(address handlerAddress, address tokenAddress, uint8 srcDecimals, uint8 destDecimals) external onlyAllowed {
+        IERCHandler handler = IERCHandler(handlerAddress);
+        handler.setDecimals(tokenAddress, srcDecimals, destDecimals);
```
Comment (lastperson) [inline contracts/Bridge.sol:171]: Source/Destination notation sounds vague, because it implies the movement from source to destination, while we are interested in tokens coming here from somewhere. So maybe we call it localDecimals and externalDecimals? Which raises a question, could there be a situation where particular token is bridged between 3 chains, and each chain has different decimals value? if yes then current solution wouldn't work, and instead we could only look at local decimals (which is already available in the token contract), while assuming that external decimals are always 18. Then whenever we emit events on the amounts, we convert them to 18 decimals?

**Review by lastperson — CHANGES_REQUESTED** [review]

**Comment by lastperson** [comment]
> > 
> 
> Deploying one contract is only 1-time cost increase. Moreover, as for me, this was the initial architectural design idea to have Handler separate so we do not bring breaking changes to the existing handler but implement new ones for new cases. cc @lastperson @mpetrun5

It is not a breaking change really, and does not require additional setup steps UNLESS there is different decimals value in different chains. Uninitialized Decimals structure should behave just like an old handler.

File: contracts/interfaces/IERCHandler.sol [hunk 5]
```diff
@@ -25,4 +25,11 @@ interface IERCHandler {
         @return address The {tokenContractAddress} that is currently set for the resourceID.
      */
     function _resourceIDToTokenContractAddress(bytes32 resourceID) external view returns (address);
+
+    /**
+      @notice Set {contractAddress} decimals on dest chain.
+      @param contractAddress Address of contract to be used when making or executing deposits.
+      @param externalDecimals Decimal places of token that is transferred.
+    */
+    function setDecimals(address contractAddress, uint8 externalDecimals) external;
```
Comment (mpetrun5) [inline contracts/interfaces/IERCHandler.sol:34]: Don't think we need this public setDecimals.
Though IERCHandler should support setResource function.

File: contracts/handlers/ERCHandlerHelpers.sol [hunk 6]
```diff
@@ -53,6 +63,16 @@ contract ERCHandlerHelpers is IERCHandler {
 
     function withdraw(bytes memory data) external virtual override {}
 
+    /**
+        @notice First verifies {contractAddress} is whitelisted,
+        then sets {_decimals}[{contractAddress}] to it's decimals value.
+        @param contractAddress Address of contract to be used when making or executing deposits.
+        @param externalDecimals Decimal places of token that is transferred.
+     */
+    function setDecimals(address contractAddress, uint8 externalDecimals) external onlyBridge {
```
Comment (mpetrun5) [inline contracts/handlers/ERCHandlerHelpers.sol:72]: Don't think we need this function.

File: migrations/6_redeploy_token_handlers.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+    let erc20HandlerInstance;
+    let xc20HandlerInstance;
+
+    // fetch deployed contracts addresses
+    try {
+      bridgeInstance = await BridgeContract.deployed();
+      erc20HandlerInstance = await ERC20HandlerContract.deployed();
+      xc20HandlerInstance = await XC20HandlerContract.deployed();
+    } catch(e){
+      console.error(e)
```
Comment (mpetrun5) [inline migrations/6_redeploy_token_handlers.js:33]: Is this okay just to print out error?
I would leave the exception as it shouldn't run anything else if it fails to fetch contract instances

Comment (nmlinaric) [reply migrations/6_redeploy_token_handlers.js:33]: If `XC20Handler` (or some other we implement in the future) isn't deployed to the current network migration an error is thrown and it stops further execution and I want it to just ignore that error and continue with other deployments/configurations. 
I was thinking about putting ERC20Handler outside try/catch since it's always deployed, and other are optional, but in the end decided to leave it as is.

Comment (mpetrun5) [reply migrations/6_redeploy_token_handlers.js:33]: Without bridge instance this all fails as well.

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -8,7 +8,8 @@ Read more [here](https://chainsafe.io/).
 
 ## Deployments
 
-To deploy contracts run `truffle migrate --network NETWORK_NAME --file <path_to_env_config>`.\
+To deploy contracts run `truffle migrate --network NETWORK_NAME --file <path_to_env_config>`.
+To deploy new handlers for tokens an register on bridge contract provide: `--redeploy-token-handlers` flag.
```
Comment (mpetrun5) [inline README.md:12]: Syntactically this seems wrong.

```suggestion
To deploy new handlers for tokens and register them on bridge contract provide: `--redeploy-token-handlers` flag.
```

**Review by mpetrun5 — CHANGES_REQUESTED** [review]

File: contracts/handlers/ERCHandlerHelpers.sol [hunk 9]
```diff
@@ -24,6 +26,14 @@ contract ERCHandlerHelpers is IERCHandler {
     // token contract address => is burnable
     mapping (address => bool) public _burnList;
 
+    // token contract address => decimals
+    mapping (address => Decimals) public _decimals;
```
Comment (lastperson) [inline contracts/handlers/ERCHandlerHelpers.sol:30]: Could we perhaps pack it together with contractWhitelist/burnList ? Or do it as a separate PR? Could leave the getters for backwards compatibility.

File: contracts/handlers/ERC20Handler.sol [hunk 10]
```diff
@@ -108,11 +110,57 @@ contract ERC20Handler is IHandler, ERCHandlerHelpers, ERC20Safe {
...
         @param resourceID ResourceID to be used when making deposits.
         @param contractAddress Address of contract to be called when a deposit is made and a deposited is executed.
         @param args Additional data to be passed to specified handler.
      */
     function setResource(bytes32 resourceID, address contractAddress, bytes calldata args) external onlyBridge {
         _setResource(resourceID, contractAddress);
+
+        uint8 externalTokenDecimals = uint8(bytes1(args));
+
+        if (externalTokenDecimals != 0) {
```
Comment (lastperson) [inline contracts/handlers/ERC20Handler.sol:123]: Should support 0 decimals as well. Need a test for 0 decimals token.

File: contracts/handlers/ERC20Handler.sol [hunk 11]
```diff
@@ -108,11 +110,57 @@ contract ERC20Handler is IHandler, ERCHandlerHelpers, ERC20Safe {
...
+        }
+    }
+
+    /**
+        @notice Converts token amount based on decimal places difference between the nework
+        deposit is made on and bridge.
+        @param tokenAddress Address of contract to be used when executing proposals.
+        @param amount Decimals value to be set for {contractAddress}.
+    */
+    function convertToExternalBalance(address tokenAddress, uint256 amount) internal returns(uint256) {
```
Comment (lastperson) [inline contracts/handlers/ERC20Handler.sol:134]: Could this be moved to the ERCHandlerHelpers?

**Review by lastperson — CHANGES_REQUESTED** [review]
Need to have a rounding loss protection logic and tests. Like when locked is 1.99 and relayed is 1.00. Or do we leave this to user discretion as it could only harm them? Possible solution is to disallow deposits that results in rounding.

File: contracts/handlers/ERCHandlerHelpers.sol (hunk 9 above)
Comment (nmlinaric) [reply contracts/handlers/ERCHandlerHelpers.sol:30]: Yep, I have a separate PR in the works for this

**Review by lastperson — APPROVED** [review]
